src/lightning/pytorch/trainer/connectors/logger_connector/result.py

```diff
@@ -411,13 +411,13 @@ def register_key(self, key: str, meta: _Metadata, value: _VALUE) -> None:
 
         Value can be provided as a nested collection
         """
-        metric = _ResultMetric(meta, isinstance(value, Tensor)).to(self.device)
```

Comment:
I think this is a good idea. In fact, let's try removing `self.device` completely and let each `_ResultMetric` use the device from each value tensor on registration.

This means something like

```python
# step 1
self.log("x", torch.tensor(1., device="cuda")

# step 2
self.log("x", torch.tensor(2., device="cpu")
```

will now fail, but this might be desirable to avoid silent correctness.